mqtt/mqtt-broker/src/persist.rs

```diff
@@ -1,28 +1,385 @@
+use std::fs::{self, OpenOptions};
+use std::io::{Read, Write};
+#[cfg(unix)]
+use std::os::unix::fs::symlink;
+#[cfg(windows)]
+use std::os::windows::fs::symlink_file;
+use std::path::PathBuf;
+use std::{cmp, fmt};
+
 use async_trait::async_trait;
+use fail::fail_point;
+use failure::ResultExt;
+use flate2::read::GzDecoder;
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use tracing::{debug, info, span, Level};
 
-use crate::error::Error;
+use crate::error::{Error, ErrorKind};
 use crate::BrokerState;
 
+/// sets the number of past states to save - 2 means we save the current and the pervious
+const STATE_DEFAULT_PREVIOUS_COUNT: usize = 2;
+static STATE_DEFAULT_STEM: &str = "state";
+static STATE_EXTENSION: &str = "dat";
+
 #[async_trait]
 pub trait Persist {
     type Error: Into<Error>;
 
-    async fn load(&mut self) -> Result<BrokerState, Self::Error>;
+    async fn load(&mut self) -> Result<Option<BrokerState>, Self::Error>;
 
     async fn store(&mut self, state: BrokerState) -> Result<(), Self::Error>;
 }
 
+/// A persistor that does nothing.
+#[derive(Debug)]
 pub struct NullPersistor;
 
 #[async_trait]
 impl Persist for NullPersistor {
     type Error = Error;
 
-    async fn load(&mut self) -> Result<BrokerState, Self::Error> {
-        Ok(BrokerState::default())
+    async fn load(&mut self) -> Result<Option<BrokerState>, Self::Error> {
+        Ok(None)
     }
 
     async fn store(&mut self, _: BrokerState) -> Result<(), Self::Error> {
         Ok(())
     }
 }
+
+/// An abstraction over the broker state's file format.
+pub trait FileFormat {
+    type Error: Into<Error>;
+
+    /// Load `BrokerState` from a reader.
+    fn load<R: Read>(&self, reader: R) -> Result<BrokerState, Self::Error>;
```

Comment:
Having used the async file operations in the past, I think it will be more straight forward to use the standard blocking calls when implementing a new `FileFormat`. This thing gets run on a completely separate thread pool with `spawn_blocking`.